### PR TITLE
New draft of power service

### PR DESCRIPTION
**Pull request by mmoskal**
formatted: https://github.com/microsoft/jacdac/blob/power_update/services/power.md

### CODE REVIEW

File: services/power.md [hunk 1]
```diff
@@ -6,62 +6,70 @@ A power-provider service.
...
+* the second word of `device_id` needs to be set to `0xAA_AA_AA_AA` (alternating 0 and 1)
+* the service index is set to `0x3d`
+* the CRC is therefore fixed
+* therefore, the packet can be recognized by reading the first 8 bytes
+
+Multi-channel power-providers (which provide a separate 900mA to each of their channels)
+will detect that a `shutdown` command is being received after 8 bytes.
+They will then briefly deactivate a switch which connects their internal and external
+Jacdac data lines.
+If the `shutdown` is still received correctly, it means it came from the internal network,
```
Comment (pelikhan) [inline services/power.md:30]: When do they send this shutdown packet packet, after closing the line?

File: services/power.md [hunk 2]
```diff
@@ -6,62 +6,70 @@ A power-provider service.
...
+The protocol is based on `shutdown` commands, which are sent as broadcast
+to the power service.
+They follow a very narrow format:
+* they need to be the only packet in the frame
+* the second word of `device_id` needs to be set to `0xAA_AA_AA_AA` (alternating 0 and 1)
+* the service index is set to `0x3d`
+* the CRC is therefore fixed
+* therefore, the packet can be recognized by reading the first 8 bytes
+
+Multi-channel power-providers (which provide a separate 900mA to each of their channels)
```
Comment (pelikhan) [inline services/power.md:26]: We should define channels since this is new terminology.

File: services/power.md [hunk 3]
```diff
@@ -6,62 +6,70 @@ A power-provider service.
...
+will detect that a `shutdown` command is being received after 8 bytes.
+They will then briefly deactivate a switch which connects their internal and external
+Jacdac data lines.
+If the `shutdown` is still received correctly, it means it came from the internal network,
+and thus there's some other power provider active.
+This makes the current power provider shut down.
+Alternatively, if the provider handles multiple channels, but has only one Jacdac input,
```
Comment (pelikhan) [inline services/power.md:33]: i'm confused by multiple channels vs 1 jacdac input.

File: services/power.md (hunk 1 above)
Comment (mmoskal) [reply services/power.md]: the shutdown command causes shut down, but clients should not listen to it; added note explaining that

File: services/power.md (hunk 2 above)
Comment (mmoskal) [reply services/power.md]: Channels defined, and I added implementation notes.

I don't believe in diagrams. Feel free to add one if you like.

File: services/power.md (hunk 3 above)
Comment (mmoskal) [reply services/power.md]: I think this is clarified now

File: services/power.md [hunk 4]
```diff
@@ -6,81 +6,203 @@ A power-provider service.
...
+Each channel can be limited to 900mA separately.
+In normal operation, the data lines of each channels are connected together.
+The ground lines are always connected together.
+Multi-channel power providers are also called _powered hubs_.
+
+While channels have separate current limits, there's nothing to prevent the user
+from joining two or more channels outside of the provider using a passive hub.
+This would allow more than 900mA of current to be drawn, resulting in cables or components
+getting hot and/or malfunctioning.
+Thus, the power negotiation protocol seeks to detect sitatations where
```
Comment (pelikhan) [inline services/power.md:27]: situations

File: services/power.md [hunk 5]
```diff
@@ -6,81 +6,203 @@ A power-provider service.
...
+multiple channels of power provider(s) are bridged together
+and shut down all but one of the channels involved.
+
+The protocol is built around the power providers periodically sending specially crafted
+`shutdown` commands in broadcast mode.
+Note that this is unusual - services typically only send reports.
+
+The `shutdown` commands can be reliably identified based on their first half.
```
Comment (pelikhan) [inline services/power.md:35]: first half of packet? more details later i guess.

File: services/power.md [hunk 6]
```diff
@@ -6,81 +6,203 @@ A power-provider service.
...
+When a power provider starts receiving a `shutdown` command, it needs to take
+steps to identify which of its channels the command is coming from.
+This is typically realized with analog switches between data lines of channels.
+The channel which received the `shutdown` command is then shut down.
+Note that in the case a single-channel provider any received `shutdown` command will cause a shut down.
+
+A multi-channel provider needs to also identify when a `shutdown` command it sent from one channel
+is received on any of its other channels and shut down one of the involved channels.
+
```
Comment (pelikhan) [inline services/power.md:44]: sketches, even hand drawn would be really helpful here. or maybe schematics

File: services/power.md [hunk 7]
```diff
@@ -6,81 +6,203 @@ A power-provider service.
...
+It is also possible to build a _data bridge_ device, with two or more ports.
+It passes through all data except for `shutdown` commands,
+but **does not** connect the power lines.
+
+### Protocol details
+
+The `shutdown` commands follow a very narrow format:
+* they need to be the only packet in the frame (and thus we can also call them `shutdown` frames)
+* the second word of `device_id` needs to be set to `0xAA_AA_AA_AA` (alternating 0 and 1)
```
Comment (pelikhan) [inline services/power.md:53]: you can now use "define" to cement those constants

**Review by pelikhan — APPROVED** [review]

**Pull request merged**